lib/sapience/extensions/grape/middleware/logging.rb

```diff
@@ -26,98 +25,56 @@ def initialize(app, options = {})
           def call!(env)
             @env = env
             before
-            error = catch(:error) do
-              begin
+
+            handle_raise do
+              @error = catch(:error) do
                 @app_response = @app.call(@env)
-              rescue StandardError => e
-                after_exception(e)
-                raise e
+                nil # no error thrown
+              end
+              if @error
+                @status = @error[:status]
+                throw(:error, @error)
               end
-              nil
-            end
-            if error
-              after_failure(error)
-              throw(:error, error)
-            else
-              @status, = *@app_response
-              after
             end
             @app_response
           end
 
-          def before
-            reset_db_runtime
-            start_time
-          end
-
-          def after
-            stop_time
-            @logger.info(parameters)
-          end
-
-          def after_exception(exc) # rubocop:disable Lint/UnusedMethodArgument
-            @status = 500
-            after
-          end
-
-          def after_failure(error)
-            @status = error[:status]
-            after
+          def handle_raise
+            begin
+              yield
+            rescue StandardError => e
+              if e.class.name =~ %r{NotFound}
```

Comment:
I see a number of issues with this approach.

This code introduces a dependency on a class name to determine the status code for the log. The class name has to match /NotFound/ regex, so the "PageNotFound" error will be logged as 404 but for instance "EntityMissing" will still be logged as 500. Not ideal.
The dependency on the class name means that when you add/change an error class, you will need to know that if you want to return 404, your error class has to be named "XyzNotFound". Basically - a developer of the app needs to know internal implementation of the sapience gem. Not ideal.
What about other errors? What if there's a validation error in the app? It will be still logged as 500 unless (using current approach) you add `elsif e.class.name =~ %r{ValidationError}`. Would you do it? I suppose you don't want to edit gems code for each error that you add in the app using that gem. Not ideal.

Eventually sapience gem is a logger and an error reporter. It should use the data that it receives from the apps using it. My suggestion in this case would be using the error object to carry data for sapience. Sapience will check if the error object responds to `code` and use it, if possible. That approach will be more flexible. Obviously it will require changing the app's code but that's why we have gem versioning - it's not mandatory to do it now.

This is my idea:

In Mark you could change this code:
https://github.com/reevoo/mark/blob/9a40f685793bc24a8174d947b604793bb7339146/config/application.rb#L21
Into this:

```ruby
class PageNotFound < RuntimeError
  attr_reader :code

  def initialize
    @code = 404
    super
  end  
end
```

Then here, in sapience you can check

```ruby
  if e.respond_to?(:code) && e.code
    @error = build_error(e, e.code)
    throw :error, @error
  else
     ...
  end
```

What do you think?

Reply:
"a developer of the app needs to know internal implementation of the sapience gem"
- Grape throws errors to generate 404s. - thats its internal implementation. - if we want to use grape we need to conform to the way it works.

"don't want to edit gems code for each error that you add in the app using that gem"
How many codes do we need? - Grape works the way it works. - On the one hand we want to raise exceptions, - and at the same time we want to add middleware logging. - these things don't go together. - if we want to `raise` we need to catch somewhere. - think about it.

"Eventually sapience gem is a logger and an error reporter"
- eventually? - no, its already a logger and an error reporter before I touched this project - You have asked that it reports the correct codes. - Whats even better than reporting the correct codes - is actually reporting what went wrong. 

~/mark/app/api/brand_reviews.rb line 13: 

`fail PageNotFound unless organisation.customer_experience_tab_enabled?`
Do we have no interest at all, in reporting to the users of our API what the actual problem was? - we just raise an exception - throw all details away, - log just `500` and the API answers with "error".  - and if we log the fact that its actually - job done, - we dont need fancy errors that actual tell people what went wrong.

Have you heard of this anti-pattern - https://stackoverflow.com/questions/729379/why-not-use-exceptions-as-regular-flow-of-control

"Exceptions are basically non-local goto statements with all the consequences of the latter. Using exceptions for flow control violates a principle of least astonishment, make programs hard to read (remember that programs are written for programmers first)."

If you want to fix the code issue here - we should fix the issue that we are using exceptions incorrectly. - from what I can see - in Gape, if we want to generate 404 - we need to raise, not throw. - do I think Grape made the right choice with this? - no I dont. - firstly - our logging middleware - is not middleware - the entire point of middleware is that it can sit in a chain of middleware with the request-response passed from one to the other. - there are many things wrong with this code, - and adding fancy exception classes - is not making it better.